### ISSUE TITLE
Supporting capistrano's clean task

### ISSUE THREAD

**Issue by denzelem** [makandra/precompiled_assets#3]
In https://github.com/makandra/precompiled_assets/blob/d2a0fc50cfa9c290284901501590d7eb8dd4bc27/lib/precompiled_assets/tasks/assets.rake#L15 we are deleting all assets, that are not part of the `manifest.js` anymore.

IMHO we should support something like https://github.com/rails/propshaft/blob/e7cdbcd4aa9c9a390dade263bbd7c7cedc0c6442/lib/propshaft/railties/assets.rake#L17 && https://github.com/rails/propshaft/blob/e7cdbcd4aa9c9a390dade263bbd7c7cedc0c6442/lib/propshaft/processor.rb#L24 in propshaft.  Then `set :keep_assets, 2` in capistrano would work as expected again, and rolling release will keep older assets available.

Wdyt?

**Comment by makmic** [comment]
Do you have any ideas on how to fix it? We currently don't have any information about previous assets in our manifests file, so we technically cannot keep X asset versions.

We could though warn if the user defines `:keep_assets` with any number except for 1 (which works as excpected)

**Comment by makmic** [comment]
one attempt would be to use something like log-rotate.
* The gem would create  a SHA of the _last_ manifest file and rename it using the SHA  (rotate)
* a new manifest is only created if assets have changed
* based on the modified timestamp of those hashed manifest files we can keep only X of them
* the contents of these manifests are kept, the rest is discarded

**Comment by denzelem** [comment]
In case the digest format is unknown within the `manifest.json`, I agree it is difficult to solve. Otherwise something similar to https://github.com/rails/propshaft/blob/e7cdbcd4aa9c9a390dade263bbd7c7cedc0c6442/lib/propshaft/output_path.rb#L10-L23C6 might work:

Example (I will use the names `logical_path` for the left part without digest and `digested_path` for the right part with digest):

```
{
  "backend.js": "backend-5dd524a6b187583cacd52e36.js",
  "cover.jpg": "cover-2b884c2c70ef803b307518e5.jpg"
}
```

1. For each `logical_path` in the manifest find all `digested_paths` in the output e.g. `public/assets`

```
backend-1d69660cb9975aa76e6f4b0e.js
backend-042c8c36f345ffa238995e82.js
backend-cf4ff9f60349dfd08cb73c2b.js
...
```

2. Sort them by `mtime`
3. Keep the newest `n` assets according to `keep_assets`
4. (Optional) Drop all other assets that are not part of the manifest anymore and are older than the `nth` oldest known asset

This might not work good with chunked assets (the have the same logical path but different digests). But it seems it is not supported by `precompiled_assets` in the current version.

**Comment by foobear** [comment]
The whole idea behind `precompiled_assets` is to not make any assumptions about how assets are built.
As long as a manifest is included, `precompiled_assets` works independently of the assets' build process.

Because of that, I believe we cannot follow the outlined approach:
* We can not know if and how digests are injected into filenames. There are lots of different ways to do it.
* Ordering assets by mtime might not always be correct. This gem has no knowledge about how and when assets were actually built.

Propshaft does not have these issues as it also controls asset generation.

The manifest is our only source of truth which is why our suggestion is to keep backups of previous manifests. (We _would_ assume that at least manifest files can be sorted reliably by mtime.)

**Comment by denzelem** [comment]
Then I'm fine with @makmic suggestion + short documentation.

> We could though warn if the user defines :keep_assets with any number except for 1 (which works as excpected)

**Comment by foobear** [comment]
Agreed, `:keep_assets` > 1 should raise an error, possibly before any deployment starts.

I believe we could still implement some kind of manifest rotation to support values above 1 as a next step.
